src/irmin/content_addressable_intf.ml

```diff
@@ -27,14 +27,19 @@ module type S = sig
   include Read_only.S
   (** @inline *)
 
+  type hash
+  (** The type for keys hashes. *)
+
+  val index : [> read ] t -> hash -> key option Lwt.t
+
   val add : [> write ] t -> value -> key Lwt.t
```

Comment:
For both`add` and `unsafe_add` I think we want to add an extra `?index:bool` argument so that users (and tests!) can decide if the want to index the hash.